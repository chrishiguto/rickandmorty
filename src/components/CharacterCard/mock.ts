const characterCardMock = {
  image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
  name: 'Rick Sanchez',
  species: 'Human'
}

export default characterCardMock
