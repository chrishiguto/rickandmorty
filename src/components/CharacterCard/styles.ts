import styled, { css, DefaultTheme } from 'styled-components'

type WrapperProps = {
  img: string
  isDead: boolean
  size: 'small' | 'large'
  hover: boolean
}

const wrapperModifiers = {
  isDead: () => css`
    filter: grayscale(1);
  `,

  small: () => css`
    height: 18.7rem;
    width: 18.7rem;
    background-position: center -3rem;
  `,

  large: () => css`
    height: 47rem;
    width: 40rem;
    background-position: center -4rem;
  `,

  hover: (theme: DefaultTheme) => css`
    &:hover,
    &:active,
    &:focus-visible {
      border-color: ${theme.colors.yellow};
      box-shadow: 0px 6px 7px ${theme.colors.yellow}25;
      outline: none;
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, img, isDead, size, hover }) => css`
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    overflow: hidden;

    background-image: url(${img});
    background-size: cover;

    border-radius: ${theme.border.radius};
    border: 0.2rem solid ${theme.colors.gray};
    backdrop-filter: blur(0);
    transition: ${theme.transition.fast};

    ${isDead && wrapperModifiers.isDead()};
    ${!!size && wrapperModifiers[size]()};
    ${hover && wrapperModifiers.hover(theme)};
  `}
`

type StripeProps = {
  transparent: boolean
}

const stripeModifiers = {
  stripe: () => css`
    background: rgba(0, 0, 0, 0.6);
    backdrop-filter: blur(2.3rem);
  `
}

export const Stripe = styled.div<StripeProps>`
  ${({ theme, transparent }) => css`
    padding: ${theme.spacings.xxsmall} 1.4rem;
    background-color: ${theme.colors.black};
    color: ${theme.colors.white};
    font-family: 'Seravek', sans-serif;
    width: 100%;
    text-align: left;

    ${transparent && stripeModifiers.stripe()}
  `}
`

export const Name = styled.h3`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xlarge};
  `}
`

export const Type = styled.p`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xsmall};
  `}
`
