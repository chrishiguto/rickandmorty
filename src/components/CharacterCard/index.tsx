import { ButtonHTMLAttributes, HTMLAttributes } from 'react'
import * as S from './styles'

type CharacterCardTypes =
  | ButtonHTMLAttributes<HTMLButtonElement>
  | HTMLAttributes<HTMLDivElement>

export type CharacterCardProps = {
  img: string
  charName: string
  charType: string
  transparentStripe?: boolean
  isDead?: boolean
  size?: 'small' | 'large'
  hover?: boolean
  as?: React.ElementType
} & CharacterCardTypes

const CharacterCard = ({
  img,
  charName,
  charType,
  transparentStripe = true,
  isDead = false,
  size = 'small',
  hover = false,
  ...props
}: CharacterCardProps) => (
  <S.Wrapper
    hover={hover}
    size={size}
    img={img}
    isDead={isDead}
    {...props}
    aria-label={`${charName} card`}
  >
    <S.Stripe
      transparent={transparentStripe}
      aria-label={`${charName}'s stripe with informations`}
    >
      <S.Name>{charName}</S.Name>
      <S.Type>{charType}</S.Type>
    </S.Stripe>
  </S.Wrapper>
)

export default CharacterCard
