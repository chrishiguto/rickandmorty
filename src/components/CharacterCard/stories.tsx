import { Story, Meta } from '@storybook/react/types-6-0'
import CharacterCard, { CharacterCardProps } from '.'

export default {
  title: 'CharacterCard',
  component: CharacterCard,
  parameters: {
    backgrounds: {
      default: 'dark'
    }
  }
} as Meta

export const Default: Story<CharacterCardProps> = (args) => (
  <CharacterCard {...args} />
)

export const asButton: Story<CharacterCardProps> = (args) => (
  <CharacterCard {...args} as="button" onClick={() => alert('Clicked')} hover />
)
