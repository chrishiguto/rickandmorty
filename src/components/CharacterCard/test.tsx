import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderWithTheme } from 'utils/tests/helpers'

import CharacterCard from '.'

const props = {
  img: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
  charName: 'Rick Sanchez',
  charType: 'Human'
}

describe('<CharacterCard />', () => {
  it('should render the character card as div by default', () => {
    const { container } = renderWithTheme(<CharacterCard {...props} />)

    expect(
      screen.getByRole('heading', { name: /rick sanchez/i })
    ).toBeInTheDocument()
    expect(screen.getByText(/human/i)).toBeInTheDocument()

    expect(screen.getByLabelText(/rick sanchez card/i)).toHaveStyleRule(
      'background-image',
      `url(${props.img})`
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the character card as button', async () => {
    const mockFn = jest.fn()

    renderWithTheme(<CharacterCard {...props} as="button" onClick={mockFn} />)

    const card = screen.getByRole('button', { name: /rick sanchez card/i })

    expect(card).toBeInTheDocument()

    userEvent.click(card)

    await waitFor(() => {
      expect(mockFn).toBeCalledTimes(1)
    })
  })

  it('should render a grayscale character card if isDead prop is true', () => {
    renderWithTheme(<CharacterCard {...props} as="button" isDead />)

    expect(screen.getByLabelText(/rick sanchez card/i)).toHaveStyleRule(
      'filter',
      'grayscale(1)'
    )
  })

  it('should render a black stripe if transparenStripe props is false', () => {
    renderWithTheme(
      <CharacterCard {...props} as="button" transparentStripe={false} />
    )

    expect(
      screen.getByLabelText(/rick sanchez's stripe with information/i)
    ).toHaveStyle({ 'background-color': '#1B1B1B' })
  })

  it('should render a large character card', () => {
    renderWithTheme(<CharacterCard {...props} as="button" size="large" />)

    expect(screen.getByLabelText(/rick sanchez card/i)).toHaveStyle({
      height: '47rem',
      width: '40rem'
    })
  })

  it('should render a yellow border and box shadow if hover prop is true', async () => {
    renderWithTheme(<CharacterCard {...props} as="button" hover />)

    const card = screen.getByLabelText(/rick sanchez card/i)

    expect(card).toHaveStyleRule('border-color', '#cbd736', {
      modifier: ':hover'
    })
  })
})
