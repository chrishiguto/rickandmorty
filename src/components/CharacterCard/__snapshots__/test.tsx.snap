// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<CharacterCard /> should render the character card as div by default 1`
] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  overflow: hidden;
  background-image: url(https://rickandmortyapi.com/api/character/avatar/1.jpeg);
  background-size: cover;
  border-radius: 0.8rem;
  border: 0.2rem solid #606060;
  -webkit-backdrop-filter: blur(0);
  backdrop-filter: blur(0);
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
  height: 18.7rem;
  width: 18.7rem;
  background-position: center -3rem;
}

.c1 {
  padding: 0.8rem 1.4rem;
  background-color: #1B1B1B;
  color: #ffffff;
  font-family: 'Seravek',sans-serif;
  width: 100%;
  text-align: left;
  background: rgba(0,0,0,0.6);
  -webkit-backdrop-filter: blur(2.3rem);
  backdrop-filter: blur(2.3rem);
}

.c2 {
  font-size: 2.0rem;
}

.c3 {
  font-size: 1.2rem;
}

<div
  aria-label="Rick Sanchez card"
  class="c0"
>
  <div
    aria-label="Rick Sanchez's stripe with informations"
    class="c1"
  >
    <h3
      class="c2"
    >
      Rick Sanchez
    </h3>
    <p
      class="c3"
    >
      Human
    </p>
  </div>
</div>
`
